content/posts/2021/11/heimdallr.md

```diff
@@ -0,0 +1,213 @@
+means that the ownership is moved into the function. This has the side effect that `buf` is no longer accessible from outside the function.
+Therefore it is impossible to modify the data buffer while the message passing operation is running. Trying to do so would lead to a compilation error.
+For a user to access the data again they need to request ownership back from the message passing operation, which happens in line 3.
+The there called `data` function of the `handle` that was returned by the non-blocking send function is an equivalent to `MPI_Wait`.
+It blocks until the used data buffer is safe to be accessed again and then returns the ownership to the caller.
+
+So in essence it is the same workflow as for an MPI application, but Rust's ownership rules allow the library to be designed in a way where
+correct and safe usage of non-blocking communication can be enforced at compile time. This is a big step up in usability and correctness
+because it is no longer on the user to remember the implicit rules of non-blocking communication but instead it is a detected program error
+if the correct procedure is not followed.
+
+This is of course just one small example on how the safety features of Rust can be used to design safer interfaces but in my opinion in showcases
+the possibilities very well.
+
+## Conclusion and further reading
+
+This blog post is supposed to give a brief overview on the challenges of message passing parallelization and how the programming interfaces used
+for it could be designed in a safer way. Parallel programming is a complex topic and introduces a variety of new error classes.
+Therefore we find it very important that the libraries and tools used for it offer as much help as possible to developers by enforcing correctness
+and detecting possible errors. 
+
+The heimdallr library introduced in this post is a prototype implementation of a message passing library that 
+concentrates on the compile time correctness aspects. It is not yet feature complete and is mainly supposed to show some of the open possibilities
+for better usability and safety in MPI.
+
+To keep this post brief I have not gone into too much detail about the implementation and some of the open problems with this solution.
+Heimdallr does have some open problems which I could not go over here without making this blog way too long. We also did not talk about
+the performance aspects, which is quite an important topic in the context of using it for HPC.
+
+If your interest was piqued a more detailed discussion about the pros and cons of heimdallr can be found in our [heimdallr paper](https://doi.org/10.1007/978-3-030-90539-2_13).
+There we also discuss some of the problems with the current implementation and show benchmark results where heimdallr's performance is compared to MPI.
+
+If you would like to try out heimdallr or have a look at the code you can visit our [github](https://github.com/parcio/heimdallr) page.
```

Comment:
- Comma after "code"
- "github" should be "GitHub"
- "repository" instead of "page"?